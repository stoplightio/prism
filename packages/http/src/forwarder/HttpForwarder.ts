import { IForwarder, IPrismInput } from '@stoplight/prism-core';
import { IHttpOperation, IServer } from '@stoplight/types';
import axios, { CancelToken } from 'axios';
import { URL } from 'url';
import { IHttpConfig, IHttpNameValue, IHttpRequest, IHttpResponse } from '../types';

export class HttpForwarder implements IForwarder<IHttpOperation, IHttpRequest, IHttpConfig, IHttpResponse> {
  public async forward(opts: {
    resource?: IHttpOperation;
    input: IPrismInput<IHttpRequest>;
    timeout?: number;
    cancelToken?: CancelToken;
  }): Promise<IHttpResponse> {
    const inputData = opts.input.data;
    const baseUrl =
      opts.resource && opts.resource.servers && opts.resource.servers.length > 0
        ? this.resolveServerUrl(opts.resource.servers[0])
        : inputData.url.baseUrl;

    if (!baseUrl) {
      throw new Error('Either one server in spec or baseUrl in request must be defined');
    }

    const response = await axios({
      method: inputData.method,
      baseURL: baseUrl,
      url: inputData.url.path,
      params: inputData.url.query,
      responseType: 'text',
      data: inputData.body,
      headers: this.updateHostHeaders(baseUrl, inputData.headers),
      validateStatus: () => true,
      timeout: Math.max(opts.timeout || 0, 0),
      ...(opts.cancelToken !== undefined && { cancelToken: opts.cancelToken }),
    });

    return {
      statusCode: response.status,
      headers: response.headers,
      body: response.data,
    };
  }

  private updateHostHeaders(baseUrl: string, headers?: IHttpNameValue) {
    // no headers? do nothing
    if (!headers) return headers;

    // host header provided? override with actual hostname
    if (headers.hasOwnProperty('host')) {
      return {
        ...headers,
        host: new URL(baseUrl).host,
        forwarded: `host=${headers.host}`,
      };
    }

    return headers;
  }

  private resolveServerUrl(server: IServer) {
    if (!server.variables) {
      return server.url;
    }

    return server.url.replace(/{(.*?)}/g, (_match, variableName) => {
      const variable = server.variables![variableName];
      if (!variable) {
        throw new Error(`Variable '${variableName}' is not defined, cannot parse input.`);
      }

      return variable.default || variable.enum![0];
    });
  }
}
