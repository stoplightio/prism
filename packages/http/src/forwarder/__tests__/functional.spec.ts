import * as axios from 'axios';

import { httpOperations } from '../../__tests__/fixtures';
import { HttpForwarder } from '../HttpForwarder';

describe('HttpForwarder', () => {
  const forwarder = new HttpForwarder();

  describe('forward()', () => {
    // todo: this test should be moved to http-server functional tests as it has nothing to do with Forwarder implementation
    it('handles multipart/form-data', async () => {
      const body =
        '--------------------------a66059a71a48e542\\r\\nContent-Disposition: form-data; name="file"; filename="file.txt"\\r\\nContent-Type: text/plain\\r\\n\\r\\FILE!\\n\\r\\n--------------------------a66059a71a48e542\r\nContent-Disposition: form-data; name="test"\r\n\r\n1\r\n--------------------------a66059a71a48e542--\r\n';
      const contentType = 'multipart/form-data; boundary=------------------------a66059a71a48e542';

      jest.spyOn(axios, 'default').mockResolvedValue({
        data: body,
        status: 200,
        statusText: 'ok',
        headers: {
          'Content-type': contentType,
        },
        config: {},
      });

      const response = await forwarder.forward({
        input: {
          validations: { input: [] },
          data: {
            method: 'post',
            url: { path: '/files' },
            headers: {
              expect: '100-continue',
              'Content-type': contentType,
              'Content-length': '286',
            },
            body,
          },
        },
        resource: Object.assign({}, httpOperations[0], {
          servers: [{ url: 'http://api.example.com' }],
        }),
      });

      expect(response).toMatchSnapshot();
    });
  });
});
