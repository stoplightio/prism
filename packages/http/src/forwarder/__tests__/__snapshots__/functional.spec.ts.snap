// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpForwarder forward() handles multipart/form-data 1`] = `
Object {
  "body": "--------------------------a66059a71a48e542\\\\r\\\\nContent-Disposition: form-data; name=\\"file\\"; filename=\\"file.txt\\"\\\\r\\\\nContent-Type: text/plain\\\\r\\\\n\\\\r\\\\FILE!\\\\n\\\\r\\\\n--------------------------a66059a71a48e542
Content-Disposition: form-data; name=\\"test\\"

1
--------------------------a66059a71a48e542--
",
  "headers": Object {
    "Content-type": "multipart/form-data; boundary=------------------------a66059a71a48e542",
  },
  "responseType": "",
  "statusCode": 200,
}
`;
